### PR TITLE
Link Yosemite to NotificationExtension which depends on it

### PR DESCRIPTION
**Pull request by mokagio**
## Description

Today, @itsmeichigo reported a CI failure p1726195899322149-slack-CC7L49W13 and identified #13896 as the first PR that, once merged, resulted in [the failure](https://buildkite.com/automattic/woocommerce-ios/builds/24609#_).

The failure:

```
ld: warning: Could not find or use auto-linked framework 'Yosemite': framework 'Yosemite' not found
❌ Undefined symbols for architecture arm64
❌   "protocol witness table for Yosemite.NotificationAction : Yosemite.Action in Yosemite", referenced from:
❌   "nominal type descriptor for Yosemite.NotificationAction", referenced from:
❌   "type metadata for Yosemite.NotificationAction", referenced from:
❌ ld: symbol(s) not found for architecture arm64
❌ clang: error: linker command failed with exit code 1 (use -v to see invocation)

[...]

** TEST BUILD FAILED **
The following build commands failed:
	Ld /opt/ci/builds/builder/automattic/woocommerce-ios/DerivedData/Build/Products/Debug-iphonesimulator/NotificationExtension.appex/NotificationExtension normal (in target 'NotificationExtension' from project 'WooCommerce')
```

After a bit of trial and error, made slow by the fact that I am not able to reproduce the issue locally (I tried all the usual things, Clean, removed `DerivedData`, run locally the same command that CI runs, ...), I noticed the following.

The changes in 1fd27830a602886f96f5bfe2b6c346e851880041 and 2e7b7e063c4d8e5446dc7d5a6962f1405ff6d2a0 add `import Yosemite` to `MarkOrderAsReadUseCase.swift`. That file is part of various targets, namely WooCommerce, WooCommerce Watch App, and NotificationExtension:

image
(_Why is it in the test target instead of being accessible there via `@testable import`?_)

WooCommerce links to Yosemite. The Watch target does not, but it's not a problem because of the `#if canImport(Yosemite)` check. But NotificationExtension did not link to Yosemite, could it then be that the `#if canImport(Yosemite)` check passed (it uses the platform destinations?) but then the build failed because the framework is not in the target?

69da800d1c4d9cbfe1c077a1720f70c99d2270a9 adds Yosemite to NotificationExtension and the build passes, which gives credence to the my guess.

The only thing that I cannot explain is why a local clean build does not run in the problem and why CI in #13896 passed and allowed the merge. My best guess is that this has something to do with Xcode parallelizing builds in a non deterministic order resulting in the framework being available during compilation (even if not directly imported) locally or in some builds but not in other. I don't know enough about the guts of the builds system to say whether this guess is solid or not.

### Where to go from here

This issue was caused by the anti pattern of sharing logic between targets by including the logic source in all targets as opposed to isolating the logic in a framework. If the logic had been isolated in a framework, the Yosemite dependency would have had to be addressed at the framework level and, once sorted, all the consumers would have access to the appropriate code without additional setup (_I think_).

I don't know the context that resulted in this design and I assume there was a reasonable tradeoff at the time to go down that path. Nevertheless, going forward I would recommend moving the this kind of shared logic in a framework or Swift package.

_If that doesn't happen, hopefully this verbose description of mine will help future versions of ourselves fix this kind of issue faster_ 😅 

For what is worth, this is something WordPress iOS was also suffering from the last time I checked...

## Steps to reproduce
N.A.

## Testing information
Since this was a CI failure, I green CI in this build should suffixe

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift [hunk 1]
```diff
@@ -18,15 +18,16 @@ struct MarkOrderAsReadUseCase {
...
         case noNeedToMarkAsRead
     }
+
+#if canImport(Yosemite)
     /// Async method that marks the order note as read if it is the notification for the last order.
     /// We do it in a way that first we syncronize notification to get the remote `Note`
     /// and then we compare local `orderID` with the one from remote `Note`.
     /// If they match we mark it as read.
     /// Returns syncronized note if marking was successful and error if some error happened
-#if canImport(Yosemite)
```
Comment (mokagio) [inline WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift:26]: Moved the header doc within the `#if` branch that conditionally defines the method the header doc documents

File: WooCommerce/WooCommerce.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -6240,13 +6240,6 @@
 			path = "Edit Product";
 			sourceTree = "<group>";
 		};
-		020C7947244E90D80098CEB2 /* Vendors */ = {
-			isa = PBXGroup;
-			children = (
-			);
-			path = Vendors;
-			sourceTree = "<group>";
-		};
```
Comment (mokagio) [inline WooCommerce/WooCommerce.xcodeproj/project.pbxproj:6249]: Dead references to an empty group.

File: fastlane/Fastfile [hunk 3]
```diff
@@ -782,25 +782,17 @@ platform :ios do
...
   end
 
-  #####################################################################################
-  # build_for_testing
-  # -----------------------------------------------------------------------------------
-  # This lane builds the app for testing
-  # -----------------------------------------------------------------------------------
-  # Usage:
-  # bundle exec fastlane build_for_testing
-  #####################################################################################
```
Comment (mokagio) [inline fastlane/Fastfile:792]: This comment didn't tell us anything that the lane name didn't already say.

File: Gemfile.lock [hunk 4]
```diff
@@ -207,7 +208,7 @@ GEM
     fastlane-plugin-appcenter (2.1.2)
     fastlane-plugin-sentry (1.24.0)
       os (~> 1.1, >= 1.1.4)
-    fastlane-plugin-wpmreleasetoolkit (12.0.0)
+    fastlane-plugin-wpmreleasetoolkit (12.1.0)
```
Comment (mokagio) [inline Gemfile.lock:211]: This is unrelated to the fix. It was just handy to upgrade to:

1. Make this available for the next code freeze
2. Not get prompted every time I run the tests locally

File: WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift [hunk 5]
```diff
@@ -18,15 +18,16 @@ struct MarkOrderAsReadUseCase {
...
     static func markOrderNoteAsReadIfNeeded(stores: StoresManager, noteID: Int64, orderID: Int) async -> Result<Note, Error> {
         let syncronizedNoteResult: Result<Note, Error> = await withCheckedContinuation { continuation in
-            let action = Yosemite.NotificationAction.synchronizeNotification(noteID: noteID) { syncronizedNote, error in
+            let action = NotificationAction.synchronizeNotification(noteID: noteID) { syncronizedNote, error in
```
Comment (mokagio) [inline WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift:30]: I thought maybe there was something odd with the namespace access that resulted in the build failure so I tried removing it. But that was not the reason.

If you find the code clear with the prefix, feel free to restore it.

File: WooCommerce/WooCommerce.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -1017,7 +1017,6 @@
 		269FFA472BF40768004E6B86 /* WooFoundationWatchOS.framework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = 269FFA452BF40768004E6B86 /* WooFoundationWatchOS.framework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
 		269FFA482BF516BA004E6B86 /* Logging.swift in Sources */ = {isa = PBXBuildFile; fileRef = 933A27362222354600C2143A /* Logging.swift */; };
 		269FFA4B2BF544C9004E6B86 /* MyStoreViewModel.swift in Sources */ = {isa = PBXBuildFile; fileRef = 269FFA4A2BF544C9004E6B86 /* MyStoreViewModel.swift */; };
-		26A000772BE9D9030071DB1E /* WooConstants.swift in Sources */ = {isa = PBXBuildFile; fileRef = B5D1AFBF20BC67C200DB0E8C /* WooConstants.swift */; };
```
Comment (mokagio) [inline WooCommerce/WooCommerce.xcodeproj/project.pbxproj:1020]: This was a duplicated source in WooFoundation. I noticed the warning in the Xcode navigator.

File: WooCommerce/WooCommerce.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -6010,6 +6011,7 @@
 				2608370B2AA66E4B0004A12B /* UserNotificationsUI.framework in Frameworks */,
 				260837092AA66E4B0004A12B /* UserNotifications.framework in Frameworks */,
 				048074D723A784C0C998FCB8 /* Pods_NotificationExtension.framework in Frameworks */,
+				3FA96DF42C94043200CDA78F /* Yosemite.framework in Frameworks */,
```
Comment (mokagio) [inline WooCommerce/WooCommerce.xcodeproj/project.pbxproj:6014]: These new `Yosemite.framework` entries are the actual fix.

File: fastlane/Fastfile [hunk 8]
```diff
@@ -782,25 +782,17 @@ platform :ios do
...
 
     run_tests(
       workspace: WORKSPACE_PATH,
       scheme: TEST_SCHEME,
       derived_data_path: 'DerivedData',
       build_for_testing: true,
-      device: options[:device],
-      deployment_target_version: options[:ios_version]
+      device: device,
+      deployment_target_version: ios_version
```
Comment (mokagio) [inline fastlane/Fastfile:795]: Converted to the new preferred named block parameters syntax while I was at it.

**Review by itsmeichigo — APPROVED** [review]
